sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -18,91 +18,79 @@
 package org.apache.beam.sdk.io.kinesis;
 
 import static com.google.common.base.Preconditions.checkNotNull;
-import static com.google.common.collect.Queues.newArrayDeque;
 
 import com.amazonaws.services.kinesis.model.ExpiredIteratorException;
 
-import java.util.Deque;
+import java.util.List;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Iterates over records in a single shard.
- * Under the hood records are retrieved from Kinesis in batches and stored in the in-memory queue.
- * Then the caller of {@link ShardRecordsIterator#next()} can read from queue one by one.
+ * Records are retrieved in batches via calls to {@link ShardRecordsIterator#readNextBatch()}.
+ * Client has to confirm processed records by calling
+ * {@link ShardRecordsIterator#ackRecord(KinesisRecord)} method.
  */
 class ShardRecordsIterator {
 
   private static final Logger LOG = LoggerFactory.getLogger(ShardRecordsIterator.class);
 
   private final SimplifiedKinesisClient kinesis;
   private final RecordFilter filter;
-  private ShardCheckpoint checkpoint;
+  private volatile ShardCheckpoint checkpoint;
```

Comment:
Not a big fan of volatiles. You could potentially use AtomicReference, but this is also quite ugly and hard to reason about.

Would it be actually possible to restructure the code a little bit, so that readNextBatch returns some kind of object, which contains both list of records and other metadata (such us "isUpToDate"?). Maybe this way you can avoid those volatiles.

Reply:
I've changed all volatiles to atomics and restructured the code a bit. I think it's now easier to reason about those shared variables.